File: src/leetcode/java/FindAllAnagramsInAString_438.java
package leetcode.java;

import java.util.LinkedList;
import java.util.List;

public class FindAllAnagramsInAString_438 {
  public List<Integer> findAnagrams(String s, String p) {
    LinkedList<Integer> res = new LinkedList<>();
    if (s == null || p == null || s.isEmpty() || p.isEmpty()) return res;

    int[] hash = new int[256];
    for (char c : p.toCharArray()) hash[c]++;
    int left = 0, right = 0, count = p.length();
    while (right < s.length()) {
      if (hash[s.charAt(right)] > 0) count--;
      hash[s.charAt(right)]--;
      right++;

      if (count == 0) res.addLast(left);

      if (right - left == p.length()) {
        if (hash[s.charAt(left)] >= 0) {
          count++;
        }
        hash[s.charAt(left)]++;
        left++;
      }
    }
    return res;
  }
}

File: src/leetcode/python/FindAllAnagramsInAString_438.py
class FindAllAnagramsInAString_438(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        res = []
        if not s or not p:
            return res
        hs = [0] * 256
        for cs in p:
            hs[ord(cs)] += 1
        l, r, c = 0, 0, len(p)
        while r < len(s):
            if hs[ord(s[r])] > 0:
                c -= 1
            hs[ord(s[r])] -= 1
            r += 1
            if c == 0:
                res.append(l)
            if r - l == len(p):
                if hs[ord(s[l])] >= 0:
                    c += 1
                hs[ord(s[l])] += 1
                l += 1
        return res
